lib/mode.coffee

```diff
@@ -0,0 +1,234 @@
+      JSON.parse message
+  .filter (m) -> m?
+  parseJsonMessages results, disabledWarnings
+
+buildMessages = (elements, disabledWarnings) ->
+  messages = []
+  lastMessage = null
+  for element in elements
+    switch element.type
+      when 'info', 'trace', 'note'
+        # Add only if there is a last message
+        if lastMessage
+          lastMessage.trace or= []
+          lastMessage.trace.push
+            type: "Trace"
+            text: element.message
+            filePath: element.file
+            range: element.range
+      when 'warning'
+        # If the message is warning and user enabled disabling warnings
+        # Check if this warning is disabled
+        if disabledWarnings and disabledWarnings.length > 0
+          messageIsDisabledLint = false
+          for disabledWarning in disabledWarnings
+            # Find a disabled lint in warning message
+            if element.message.indexOf(disabledWarning) >= 0
+              messageIsDisabledLint = true
+              lastMessage = null
+              break
+          if not messageIsDisabledLint
+            lastMessage = constructMessage "Warning", element
+            messages.push lastMessage
+        else
+          lastMessage = constructMessage "Warning" , element
+          messages.push lastMessage
+      when 'error', 'fatal error'
+        lastMessage = constructMessage "Error", element
+        messages.push lastMessage
+  return messages
+
+constructMessage = (type, element) ->
+  message =
+    type: type
+    text: element.message
+    filePath: element.file
+    range: element.range
+  # children exists only in JSON messages
+  if element.children
+    message.trace = []
+    for children in element.children
+      message.trace.push
+        type: "Trace"
+        text: children.message
+        filePath: element.file
+        range: children.range or element.range
+  message
+
+buildRustcArguments = (linter, paths) ->
+  [editingFile, cargoManifestPath] = paths
+  Promise.resolve().then () =>
+    rustcArgs = switch linter.rustcBuildTest
+      when true then ['--cfg', 'test', '-Z', 'no-trans', '--color', 'never']
+      else ['-Z', 'no-trans', '--color', 'never']
+    cmd = [linter.rustcPath]
+      .concat rustcArgs
+    if cargoManifestPath
+      cmd.push '-L'
+      cmd.push path.join path.dirname(cargoManifestPath), linter.cargoDependencyDir
+    compilationFeatures = linter.compilationFeatures(false)
+    cmd = cmd.concat compilationFeatures if compilationFeatures
+    cmd = cmd.concat [editingFile]
+    [editingFile, cmd]
+
+buildCargoArguments = (linter, cargoManifestPath) ->
+  cargoArgs = switch linter.cargoCommand
+    when 'check' then ['check']
+    when 'test' then ['test', '--no-run']
+    when 'rustc' then ['rustc', '-Zno-trans', '--color', 'never']
+    when 'clippy' then ['clippy']
+    else ['build']
+
+  compilationFeatures = linter.compilationFeatures(true)
+  buildCargoPath(linter.cargoPath).then (cmd) ->
```

Comment:
🐛 I think you missed a `linter.cargoCommand` as a second argument to this function. As it is currently written since you aren't providing a `cargoCommand` here, this will _always_ return `[cargoPath]`, so might as well just remove the entire `buildCargoPath` function as this is the only location it is being called from.

If the `buildCargoPath` function is being removed, so should the `usingMultitoolForClippy` as it is only used in there.